src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicEditorNavigationBarItemService.vb

```diff
@@ -40,41 +41,32 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.NavigationBar
-            Dim location As Location = Await GetSourceNavigationLocationAsync(document, symbol, cancellationToken).ConfigureAwait(False)
-            If location Is Nothing Then
-                Return Nothing
-            End If
-
-            ' If the symbol is a method symbol, we'll figure out the right location which may be in
-            ' virtual space
-            If symbol.Kind = SymbolKind.Method Then
-                Dim methodBlock = location.FindToken(cancellationToken).GetAncestor(Of MethodBlockBaseSyntax)()
+        Friend Overrides Async Function GetNavigationLocationAsync(
+                document As Document,
+                item As NavigationBarItem,
+                symbolItem As SymbolItem,
+                textSnapshot As ITextSnapshot,
+                cancellationToken As CancellationToken) As Task(Of (documentId As DocumentId, position As Integer, virtualSpace As Integer))
 
-                If methodBlock IsNot Nothing Then
-                    Return NavigationPointHelpers.GetNavigationPoint(location.SourceTree.GetText(cancellationToken), 4, methodBlock)
-                End If
-            End If
-
-            Return New VirtualTreePoint(location.SourceTree, location.SourceTree.GetText(cancellationToken), location.SourceSpan.Start)
-        End Function
+            Dim navigationLocation = Await MyBase.GetNavigationLocationAsync(document, item, symbolItem, textSnapshot, cancellationToken).ConfigureAwait(False)
 
-        Private Shared Async Function GetSourceNavigationLocationAsync(document As Document, symbol As ISymbol, cancellationToken As CancellationToken) As Task(Of Location)
-            Dim sourceLocations = symbol.Locations.Where(Function(l) l.IsInSource)
+            Dim destinationDocument = document.Project.Solution.GetDocument(navigationLocation.documentId)
 
-            ' First figure out the location that we want to grab considering partial types
-            Dim syntaxTree = Await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(False)
-            Dim location = sourceLocations.FirstOrDefault(Function(l) l.SourceTree.Equals(syntaxTree))
+            Dim root = Await destinationDocument.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(False)
 
-            If location Is Nothing Then
-                location = sourceLocations.FirstOrDefault
+            ' If the symbol is a method symbol, we'll figure out the right location which may be in virtual space
+            Dim methodBlock = root.FindToken(navigationLocation.position).GetAncestor(Of MethodBlockBaseSyntax)()
+            If methodBlock IsNot Nothing Then
+                Dim text = Await destinationDocument.GetTextAsync(cancellationToken).ConfigureAwait(False)
+                Dim navPoint = NavigationPointHelpers.GetNavigationPoint(text, indentSize:=4, methodBlock)
+                Return (navigationLocation.documentId, navPoint.Position, navPoint.VirtualSpaces)
             End If
```

Comment:
this part is still unfortunate.  VB has some really funky legacy behavior here around jumping to a blank line (including into virtual space).  I just didn't want to clutter the underlying api even more with this, so i kept this part at the vb editor level.  Note though that at least now we still avoid semantics.  All the work done here just uses syntax/text, so it should still be better than before.